x/ccv/provider/keeper/keeper.go

```diff
@@ -1184,3 +1184,141 @@ func (k Keeper) IsOptIn(ctx sdk.Context, chainID string) bool {
+
+	store.Set(types.OptedInKey(chainID, providerAddr), blockHeightBytes)
+}
+
+func (k Keeper) DeleteOptedIn(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.OptedInKey(chainID, providerAddr))
+}
+
+func (k Keeper) IsOptedIn(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Get(types.OptedInKey(chainID, providerAddr)) != nil
+}
+
+func (k Keeper) GetOptedIn(
+	ctx sdk.Context,
+	chainID string) (optedInValidators []OptedInValidator) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.ChainIdWithLenKey(types.OptedInBytePrefix, chainID)
+	iterator := sdk.KVStorePrefixIterator(store, key)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		optedInValidators = append(optedInValidators, OptedInValidator{
+			ProviderAddr: types.NewProviderConsAddress(iterator.Key()[len(key):]),
+			BlockHeight:  binary.BigEndian.Uint64(iterator.Value()),
+		})
+	}
+
+	return optedInValidators
+}
+
+func (k Keeper) SetToBeOptedIn(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.ToBeOptedInKey(chainID, providerAddr), []byte{})
+}
+
+func (k Keeper) DeleteToBeOptedIn(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.ToBeOptedInKey(chainID, providerAddr))
+}
+
+func (k Keeper) IsToBeOptedIn(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Get(types.ToBeOptedInKey(chainID, providerAddr)) != nil
+}
+
+func (k Keeper) GetToBeOptedIn(
+	ctx sdk.Context,
+	chainID string) (addresses []types.ProviderConsAddress) {
+
+	store := ctx.KVStore(k.storeKey)
+	key := types.ChainIdWithLenKey(types.ToBeOptedInBytePrefix, chainID)
+	iterator := sdk.KVStorePrefixIterator(store, key)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		providerAddr := types.NewProviderConsAddress(iterator.Key()[len(key):])
+		addresses = append(addresses, providerAddr)
+	}
+
+	return addresses
+}
+
+func (k Keeper) SetToBeOptedOut(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.ToBeOptedOutKey(chainID, providerAddr), []byte{})
+}
+
+func (k Keeper) DeleteToBeOptedOut(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.ToBeOptedOutKey(chainID, providerAddr))
+}
+
+func (k Keeper) IsToBeOptedOut(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Get(types.ToBeOptedOutKey(chainID, providerAddr)) != nil
+}
+
+func (k Keeper) GetToBeOptedOut(
```

Comment:
Might just be me, but the modifications in this file seem like tripley duplicated code,
the only thing that really changes is the Key/Preifx. Feel free to ignore this comment, but
it might be worth refactoring to make this more concise (probably at the expense of adding a layer of abstraction)

Reply:
That's true. At least the `ToBeOptedIn` and `ToBeOptedOut` are pretty much the same but with a different key. Would think if there is a nice way to refactor this by just providing the key in a future iteration.